File: web/lib/handcrop.ts
/**
 * MediaPipe hand detection + crop — the real-world robustness keystone.
 *
 * Before classifying, we detect the hand and crop to a normalized square box
 * around it. This removes most of the background dependence that makes the
 * model (trained on uniform images) struggle on a cluttered webcam.
 *
 * This is the browser half of a cropping step that must also exist in the
 * Python pipeline for training/eval parity (see the ML robustness workstream).
 * The crop geometry here (square box, margin fraction) is the contract both
 * sides share.
 *
 * Runs 100% client-side via MediaPipe Tasks Vision (WASM). The model and WASM
 * assets are served from the site's own /mediapipe/ path so there is no runtime
 * CDN dependency and frames never leave the browser.
 */

import {
  HandLandmarker,
  FilesetResolver,
  type HandLandmarkerResult,
} from "@mediapipe/tasks-vision";

/** A normalized crop box in [0,1] image coordinates. */
export interface CropBox {
  x: number;
  y: number;
  width: number;
  height: number;
}

/** Result of a hand-detection pass. */
export interface HandDetection {
  /** True if at least one hand was found. */
  found: boolean;
  /** The square crop box (normalized), present only when {@link found}. */
  box?: CropBox;
}

/**
 * Fraction of the hand's bounding-box size added as margin on each side, so the
 * crop includes the whole hand plus context (matches the training crop policy).
 */
export const CROP_MARGIN = 0.35;

const WASM_PATH = "/mediapipe/wasm";
const MODEL_PATH = "/mediapipe/hand_landmarker.task";

let landmarkerPromise: Promise<HandLandmarker> | null = null;

/**
 * Lazily create (once) the shared {@link HandLandmarker}, configured for video.
 *
 * @returns A promise resolving to the cached landmarker.
 */
export function getHandLandmarker(): Promise<HandLandmarker> {
  if (!landmarkerPromise) {
    landmarkerPromise = (async () => {
      const fileset = await FilesetResolver.forVisionTasks(WASM_PATH);
      return HandLandmarker.createFromOptions(fileset, {
        baseOptions: { modelAssetPath: MODEL_PATH },
        runningMode: "VIDEO",
        numHands: 1,
      });
    })();
  }
  return landmarkerPromise;
}

/** Reset the cached landmarker (used by tests). */
export function resetHandLandmarker(): void {
  landmarkerPromise = null;
}

let imageLandmarkerPromise: Promise<HandLandmarker> | null = null;

/**
 * Lazily create (once) a SEPARATE {@link HandLandmarker} configured for stills.
 *
 * The shared {@link getHandLandmarker} instance runs in `"VIDEO"` mode and can
 * only be driven via `detectForVideo`; calling `.detect()` on it throws. Uploads
 * are single still images, so they need their own `"IMAGE"`-mode landmarker.
 * Both share the same WASM fileset and model asset.
 *
 * @returns A promise resolving to the cached image-mode landmarker.
 */
export function getImageHandLandmarker(): Promise<HandLandmarker> {
  if (!imageLandmarkerPromise) {
    imageLandmarkerPromise = (async () => {
      const fileset = await FilesetResolver.forVisionTasks(WASM_PATH);
      return HandLandmarker.createFromOptions(fileset, {
        baseOptions: { modelAssetPath: MODEL_PATH },
        runningMode: "IMAGE",
        numHands: 1,
      });
    })();
  }
  return imageLandmarkerPromise;
}

/** Reset the cached image-mode landmarker (used by tests). */
export function resetImageHandLandmarker(): void {
  imageLandmarkerPromise = null;
}

/**
 * Detect a hand in a single still image and return its normalized crop box.
 *
 * Uses the `"IMAGE"`-mode landmarker (so `.detect()` is valid). Returns the same
 * {@link HandDetection} shape as {@link cropBoxFromLandmarks}: `found:false` when
 * no hand is present, so callers can fall back to whole-image classification.
 *
 * @param image - A decoded image (or other still source) to detect a hand in.
 * @returns A {@link HandDetection} with the square crop box when a hand is found.
 */
export async function detectHandInImage(
  image: ImageBitmapSource | HTMLImageElement | HTMLCanvasElement,
): Promise<HandDetection> {
  const landmarker = await getImageHandLandmarker();
  const result = landmarker.detect(image as Parameters<HandLandmarker["detect"]>[0]);
  return cropBoxFromLandmarks(result);
}

/**
 * Compute a square, margined crop box from MediaPipe landmark results.
 *
 * Pure and dependency-free so it can be unit-tested without the WASM runtime:
 * pass a {@link HandLandmarkerResult}-shaped object and assert the box. The box
 * is the landmarks' bounding square expanded by {@link CROP_MARGIN}, clamped to
 * the unit image.
 *
 * @param result - The landmarker output (uses `result.landmarks[0]`).
 * @param margin - Margin fraction (defaults to {@link CROP_MARGIN}).
 * @returns A {@link HandDetection}; `found:false` when no hand is present.
 */
export function cropBoxFromLandmarks(
  result: Pick<HandLandmarkerResult, "landmarks">,
  margin: number = CROP_MARGIN,
): HandDetection {
  const hand = result.landmarks?.[0];
  if (!hand || hand.length === 0) {
    return { found: false };
  }

  let minX = Infinity;
  let minY = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;
  for (const point of hand) {
    if (point.x < minX) minX = point.x;
    if (point.y < minY) minY = point.y;
    if (point.x > maxX) maxX = point.x;
    if (point.y > maxY) maxY = point.y;
  }

  const w = maxX - minX;
  const h = maxY - minY;
  // Square side = larger dimension, expanded by margin on both sides.
  const side = Math.max(w, h) * (1 + 2 * margin);
  const cx = (minX + maxX) / 2;
  const cy = (minY + maxY) / 2;

  let x = cx - side / 2;
  let y = cy - side / 2;
  let boxSide = side;
  // Clamp to the unit image while keeping the box square.
  if (x < 0) x = 0;
  if (y < 0) y = 0;
  if (x + boxSide > 1) boxSide = Math.min(boxSide, 1 - x);
  if (y + boxSide > 1) boxSide = Math.min(boxSide, 1 - y);

  return { found: true, box: { x, y, width: boxSide, height: boxSide } };
}

/**
 * Draw the region of a source defined by a normalized {@link CropBox} into a
 * fresh canvas of `outSize x outSize` pixels (square), ready for preprocessing.
 *
 * @param source - The video/image/canvas to crop from.
 * @param sourceWidth - Source pixel width.
 * @param sourceHeight - Source pixel height.
 * @param box - Normalized crop box.
 * @param outSize - Output canvas side length.
 * @returns The cropped square canvas.
 */
export function cropToCanvas(
  source: CanvasImageSource,
  sourceWidth: number,
  sourceHeight: number,
  box: CropBox,
  outSize: number,
): HTMLCanvasElement {
  const canvas = document.createElement("canvas");
  canvas.width = outSize;
  canvas.height = outSize;
  const ctx = canvas.getContext("2d");
  if (!ctx) throw new Error("Could not acquire a 2D context for cropping.");
  ctx.drawImage(
    source,
    box.x * sourceWidth,
    box.y * sourceHeight,
    box.width * sourceWidth,
    box.height * sourceHeight,
    0,
    0,
    outSize,
    outSize,
  );
  return canvas;
}

File: src/handcrop.py
"""MediaPipe hand detection + crop — the Python half of the crop contract.

This mirrors the browser implementation in ``web/lib/handcrop.ts`` so the crop
geometry is identical on both sides: detect the hand, take the landmarks'
bounding square, expand it by :data:`CROP_MARGIN` on each side, clamp to the
image, and crop. Cropping to the hand before classifying removes most of the
background dependence that hurts the model (trained on uniform images) on
cluttered real-world photos.

The crop math (:func:`crop_box_from_landmarks`) is pure and unit-tested without
the MediaPipe runtime; :func:`detect_and_crop` adds the actual detection. The
``mediapipe`` import is deferred so importing this module (and the rest of the
pipeline) does not hard-require the heavy dependency.

The shared geometry contract — square box, :data:`CROP_MARGIN` fraction — MUST
stay in sync with ``web/lib/handcrop.ts`` (``CROP_MARGIN`` there).
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

import numpy as np
from PIL import Image

# Fraction of the hand's bounding-box size added as margin on each side. Must
# match CROP_MARGIN in web/lib/handcrop.ts.
CROP_MARGIN = 0.35

# Default location of the MediaPipe hand-landmarker model (shared with the web
# app, which serves it as a static asset).
DEFAULT_MODEL_PATH = "web/public/mediapipe/hand_landmarker.task"


@dataclass(frozen=True)
class CropBox:
    """A normalized crop box in ``[0, 1]`` image coordinates."""

    x: float
    y: float
    width: float
    height: float


def crop_box_from_landmarks(
    landmarks: list[tuple[float, float]] | np.ndarray,
    margin: float = CROP_MARGIN,
) -> CropBox | None:
    """Compute a square, margined crop box from normalized hand landmarks.

    Pure and dependency-free (no MediaPipe needed) so it is directly testable.
    Returns ``None`` when there are no landmarks.

    Args:
        landmarks: Sequence of ``(x, y)`` points in ``[0, 1]`` image coords.
        margin: Margin fraction (defaults to :data:`CROP_MARGIN`).

    Returns:
        A square :class:`CropBox` clamped to the unit image, or ``None``.
    """
    pts = np.asarray(landmarks, dtype=np.float64)
    if pts.size == 0:
        return None

    min_x, min_y = pts[:, 0].min(), pts[:, 1].min()
    max_x, max_y = pts[:, 0].max(), pts[:, 1].max()

    w = max_x - min_x
    h = max_y - min_y
    side = max(w, h) * (1 + 2 * margin)
    cx = (min_x + max_x) / 2
    cy = (min_y + max_y) / 2

    x = max(0.0, cx - side / 2)
    y = max(0.0, cy - side / 2)
    box_side = side
    if x + box_side > 1:
        box_side = min(box_side, 1 - x)
    if y + box_side > 1:
        box_side = min(box_side, 1 - y)

    return CropBox(
        x=float(x), y=float(y), width=float(box_side), height=float(box_side)
    )


def crop_image(image: Image.Image, box: CropBox) -> Image.Image:
    """Crop a PIL image to a normalized :class:`CropBox` (returns the region)."""
    w, h = image.size
    left = int(round(box.x * w))
    top = int(round(box.y * h))
    right = int(round((box.x + box.width) * w))
    bottom = int(round((box.y + box.height) * h))
    # Guard against a degenerate zero-area box.
    right = max(right, left + 1)
    bottom = max(bottom, top + 1)
    return image.crop((left, top, right, bottom))


def _build_landmarker(model_path: str | Path) -> Any:
    """Create a MediaPipe HandLandmarker for still images (deferred import)."""
    from mediapipe.tasks import python
    from mediapipe.tasks.python import vision

    base_options = python.BaseOptions(model_asset_path=str(model_path))
    options = vision.HandLandmarkerOptions(
        base_options=base_options,
        running_mode=vision.RunningMode.IMAGE,
        num_hands=1,
    )
    return vision.HandLandmarker.create_from_options(options)


def detect_and_crop(
    image: Image.Image,
    landmarker: Any | None = None,
    model_path: str | Path = DEFAULT_MODEL_PATH,
    margin: float = CROP_MARGIN,
) -> Image.Image | None:
    """Detect the hand in ``image`` and return the cropped region, or ``None``.

    Args:
        image: An RGB PIL image.
        landmarker: A pre-built MediaPipe HandLandmarker to reuse across calls;
            if ``None``, one is created from ``model_path`` for this call.
        model_path: Path to ``hand_landmarker.task`` (when ``landmarker`` is None).
        margin: Crop margin fraction.

    Returns:
        The cropped hand region as a PIL image, or ``None`` if no hand is found.
    """
    import mediapipe as mp

    own = landmarker is None
    if landmarker is None:
        landmarker = _build_landmarker(model_path)
    try:
        rgb = np.asarray(image.convert("RGB"), dtype=np.uint8)
        mp_image = mp.Image(image_format=mp.ImageFormat.SRGB, data=rgb)
        result = landmarker.detect(mp_image)
        if not result.hand_landmarks:
            return None
        pts = [(lm.x, lm.y) for lm in result.hand_landmarks[0]]
        box = crop_box_from_landmarks(pts, margin=margin)
        if box is None:
            return None
        return crop_image(image, box)
    finally:
        if own:
            landmarker.close()
